xla/service/gpu/double_buffer_loop_unrolling.cc

```diff
@@ -83,6 +83,214 @@ void SetChannelIdForNewCollective(HloInstruction* new_instr,
+                        HloOpcode::kRecv>(old_instr)) {
+    return absl::OkStatus();
+  }
+  const google::protobuf::Map<std::string, std::string>& attribute_map =
+      new_instr->frontend_attributes().map();
+  if (!attribute_map.contains(kSendRecvValidationAttr)) {
+    return absl::OkStatus();
+  }
+  VLOG(3) << "Original send-recv iterations: "
+          << attribute_map.at(kSendRecvValidationAttr);
+
+  TF_ASSIGN_OR_RETURN(
+      auto send_recv_validation_attr,
+      ParseVectorOfPairs(attribute_map.at(kSendRecvValidationAttr)));
+
+  if (send_recv_validation_attr.size() == 0) {
+    return absl::OkStatus();
+  }
+
+  std::vector<Interval> send_recv_iterations_new_instr1,
+      send_recv_iterations_new_instr2;
+  send_recv_iterations_new_instr1.reserve(send_recv_validation_attr.size());
+  send_recv_iterations_new_instr2.reserve(send_recv_validation_attr.size());
+  for (const Interval& pair : send_recv_validation_attr) {
+    int64_t a = pair.first;
+    int64_t b = pair.second;
+    if (is_peeled) {
+      send_recv_iterations_new_instr1.emplace_back(
+          std::floor(a / 2.0), std::max(0.0, std::floor((b - 1) / 2.0)));
+      send_recv_iterations_new_instr2.emplace_back(
+          std::max(0.0, std::floor((a - 1) / 2.0)),
+          std::max(0.0, std::floor((b - 2) / 2.0)));
+    } else {
+      send_recv_iterations_new_instr1.emplace_back(std::floor((a + 1) / 2.0),
+                                                   std::floor(b / 2.0));
+      send_recv_iterations_new_instr2.emplace_back(
+          std::floor(a / 2.0), std::max(0.0, std::floor((b - 1) / 2.0)));
+    }
+  }
+
+  std::string iteration_instances_new_instr1 =
+      "{" +
+      absl::StrJoin(send_recv_iterations_new_instr1, ",",
+                    [](std::string* out, Interval item) {
+                      absl::StrAppend(out, "{", item.first, ",", item.second,
+                                      "}");
+                    }) +
+      "}";
+  std::string iteration_instances_new_instr2 =
+      "{" +
+      absl::StrJoin(send_recv_iterations_new_instr2, ",",
+                    [](std::string* out, Interval item) {
+                      absl::StrAppend(out, "{", item.first, ",", item.second,
+                                      "}");
+                    }) +
+      "}";
+
+  VLOG(3) << "New send-recv iterations for original instruction: "
+          << iteration_instances_new_instr1;
+  VLOG(3) << "New send-recv iterations for new instruction: "
+          << iteration_instances_new_instr2;
+
+  FrontendAttributes attributes;
+  // Copy the attributes from old_instr attributes
+  for (const auto& [key, val] : old_instr->frontend_attributes().map()) {
+    if (key == kSendRecvValidationAttr) {
+      continue;
+    }
+    (*attributes.mutable_map())[key] = val;
+  }
+  (*attributes.mutable_map())[kSendRecvValidationAttr] =
+      iteration_instances_new_instr1;
+  old_instr->set_frontend_attributes(attributes);
+
+  // Do the same for new instr
```

Comment:
This looks similar. Can you extraxt this into a function `CopyAttributes` with a skip_key param. 
Or something like that..